### PR TITLE
Implement ListOfVariableReferences

### PR DESCRIPTION
**Pull request by odow**
<s>Not tested yet as master on MOI and MOIU don't play nicely and I need to find a commit that does.</s>

Closes https://github.com/JuliaOpt/MathOptInterfaceUtilities.jl/issues/25

### CODE REVIEW

File: src/instance.jl [hunk 1]
```diff
@@ -49,7 +49,11 @@ getconstrloc(m::AbstractInstance, cr::CR) = m.constrmap[cr.value]
 
 # Variables
 MOI.get(m::AbstractInstance, ::MOI.NumberOfVariables) = m.nvars
```
Comment (odow) [inline src/instance.jl]: `length(m.varrefs)` as this doesn't account for deleted variables?

File: src/instance.jl [hunk 2]
```diff
@@ -89,16 +93,18 @@ function MOI.delete!(m::AbstractInstance, vr::MOI.VariableReference)
...
     end
+    splice!(m.varrefs, vf)
     if haskey(m.varnames, vr.value)
         delete!(m.namesvar, m.varnames[vr.value])
         delete!(m.varnames, vr.value)
     end
 end
 
 MOI.isvalid(m::AbstractInstance, cr::MOI.ConstraintReference) = !iszero(m.constrmap[cr.value])
+MOI.isvalid(m::AbstractInstance, vr::MOI.VariableReference) = in(vf, m.varrefs)
```
Comment (blegat) [inline src/instance.jl]: Wouldn't this be extremely inefficient if there are many variables ? I think we should rather use a `Set` than a `Vector`

File: src/instance.jl [hunk 3]
```diff
@@ -89,16 +93,18 @@ function MOI.delete!(m::AbstractInstance, vr::MOI.VariableReference)
     for cr in rm
         MOI.delete!(m, cr)
     end
+    splice!(m.varrefs, vf)
```
Comment (blegat) [inline src/instance.jl]: The second argument of `splice!` should be the indices. What is `vf` ? I think `delete!(m.varrefs, vr)` would work if `vr` is a `Set`.

File: src/instance.jl (hunk 1 above)
Comment (blegat) [reply src/instance.jl]: Yes, it should be `length(m.varrefs)`, `m.nvars` is incorrect

Comment (blegat) [reply src/instance.jl]: By the way `nvars` (resp. `nconstrs`) should rather be named `curvarid` or `nextvarid`

File: src/instance.jl (hunk 2 above)
Comment (odow) [reply src/instance.jl]: Probably. But it's not clear if that will make much of a difference in the grand scheme of things. Depends how often people plan on checking whether a VariableReference is valid.

**Pull request merged**